Components/jason/riscv fs

powerjg opened the pull request.

powerjg committed:
configs: Add RISC-V board to components

Change-Id: Ie098e7cd94c8f8b5b4036a478ee8166b1fb9b263
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs: Add RISC-V FS example to components

Change-Id: Ib4fe99de7a1fe82c787a8c38d36bf7c7f5cb277c
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs: Create interrupts for ISAs other than x86

All CPUs need to have `createInterrupts()` called. Add a switch to check
for that in the caches and make the ports optional parameters.

Change-Id: I38310c57a68ef18fbe1c28844dcda515eca3170e

powerjg committed:
configs: Remove KVM from default imports in components

Change-Id: I74028df1f775564ff16a41ab367ddad512db524c
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs: Expose the MMU to the board in components

Change-Id: I30b4736ebe44b8429a32c8951af6e654a1238ae6
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
arch-riscv,dev: Make DTB not depend on python name

Currently, the RISC-V device tree generation assumes that you have very
specific variable names for the instantiated SimObjects in your config
file. For instance, it assumes that `system.cpu` is a list of BaseCPUs.
This change instead uses the state that's passed around with the device
tree generation to figure out how many CPUs there should be and uses a
unique string as the key instead of the SimObject id.

Change-Id: I4a7c670428e7068862f991577787644a8f91c72c
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
arch-riscv,dev: Explicitly set num CPUs on platfrom

Previously, the RISC-V devices queried the system object in
SimObject::init() for the number of CPUs and the number of threads.
However, the system object doesn't actually count the number of
CPUs/threads until it runs init(). Therefore, we've just been getting
lucky in the order that the SimObject init() functions were called.

This change instead decouples these two functions and makes the number
of CPUs/threads a parameter for the RISC-V interrupt devices. This
change also updates the example config script.

Change-Id: Ic4da5604156837cfeec05e58d188b42a02420de1
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

BobbyRBruce commented on components_library/boards/riscv_board.py:
```diff
@@ -0,0 +1,280 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from configs.common.Caches import IOCache
```
I really don't like relying on anything in `configs` inside the component library. While stuff in configs can use the components library, I don't think otherway around should be permitted. If this is needed, could we migrate it?

powerjg replied:
Whoa, this was a big mistake on my part! I blame VS code for auto importing it :)

BobbyRBruce commented on components_library/boards/riscv_board.py:
```diff
@@ -0,0 +1,280 @@
+import os
+from typing import Optional
+
+from components_library.utils.override import overrides
```
It's not a big deal, but in general I've been using relative paths as much as possible within the library.

BobbyRBruce commented on components_library/boards/riscv_board.py:
```diff
@@ -0,0 +1,280 @@
+        fdt.writeDtbFile(os.path.join(outdir, "device.dtb"))
+
+    # TODO: Finish implementing this.
+    def generateDeviceTree(self, state):
```
This function is never called, right?

powerjg replied:
Correct. Hence the TODO

BobbyRBruce commented on components_library/cachehierarchies/classic/no_cache.py:
```diff
@@ -113,6 +113,8 @@ def incorporate_cache(self, board: AbstractBoard) -> None:
                 int_resp_port = self.membus.cpu_side_ports
                 core.connect_interrupt(int_req_port, int_resp_port)
+            else:
+                core.connect_interrupt()
```
Just to clarify, this holds true for all non-X86 targets?

powerjg replied:
It's definitely true that all targets must have `BaseCPU.createInterrupts()` called on it. Whether this is the right place to do it or not is up for debate :)

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+
+## General components improvements
+
+- Add a "CompatibilityError" instead of "EnvironmentError"
```
Yes, i think someone should sit down and create some bespoke gem5 Errors. I've not been thinking very hard about this and generally throwing rather generic exceptions. There's a Jira ticket for this.

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+## General components improvements
+
+- Add a "CompatibilityError" instead of "EnvironmentError"
+- make `is_ruby` a member of the cache hierarchy, not an "environment" variable.
```
It is isn't it? It's an abtract function in AbstractCacheHierarchy.

powerjg replied:
Yeah, I found that after making this note. That's not how the x86 board does it though.

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+- Add a "CompatibilityError" instead of "EnvironmentError"
+- make `is_ruby` a member of the cache hierarchy, not an "environment" variable.
+- Add a helper for disk images?
+- Limitations of SimpleBoard doesn't seem true anymore
```
If we can support full system on the simple board, that'd be great!

powerjg replied:
I see what you're saying now. The `SimpleBoard` only supports SE mode by itself.

What I am saying is that things that derive from `SimpleBoard` support FS. I think the documentation is a bit confusing. I am on the fence as to whether we should have a separate SE vs FS base board.

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+- make `is_ruby` a member of the cache hierarchy, not an "environment" variable.
+- Add a helper for disk images?
+- Limitations of SimpleBoard doesn't seem true anymore
+- When can I use the `get_processor` and `get_cores` functions?
```
Whenever you want. As soon as these functions are available they are usable. The processor exists within a board as soon as you construct the board. The processor contains the cores as soon as its constructed. `incorporate_processor` is simply connecting things to the board correctly.

powerjg replied:
We should document this :)

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+
+- What's the point of SimpleBoard and why does X86Board inherit it?
+  - Has some useful functions... document this
+- Where do I get the memory size? Is this a parameter memory that the system should grab, or is it a parameter that system should set on the memory?
```
board.get_memory().get_memory_ranges()

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+
+## Things that could be useful
+
+A `require` function which, for instance, requires that ISA is RISCV.
```
Yes, I think a more clever way to delcare run-time requirements would be useful and remove a lot of boilerplate. For example, when I write a config script I'm having to put a lot of if-statements at the top making sure everything is ok, otherwise return an error (e.g., "You need to have gem5 compiled to X86").

BobbyRBruce commented on components_library/notes.md:
```diff
@@ -0,0 +1,62 @@
+# Notes setting up RISC-V FS in components
```
Lots of good stuff here, I'll add some of the more interesting ones as Jira issues this afternoon.

BobbyRBruce commented on components_library/processors/simple_core.py:
```diff
@@ -95,3 +97,7 @@ def connect_interrupt(
             self.core.interrupts[0].pio = interrupt_requestor
             self.core.interrupts[0].int_requestor = interrupt_responce
             self.core.interrupts[0].int_responder = interrupt_requestor
```
Does gem5 work ok if you set these to None?

powerjg replied:
Hmm, probably not. They are required when x86 and not required otherwise. Maybe instead the condition should be `if interrupt_requestor or interrupt_responder` or something like that instead of checking for x86 multiple places.

BobbyRBruce replied:
Don't forget to implement this :)

powerjg replied:
Do I need to do something more right now? Currently, it's checking to see if it's x86. We could probably clean that up, but it's not currently in my scope. I have a few other open things in this PR that I'll try to come back to once I've pushed this relation chain upstream.

BobbyRBruce replied:
Ok. It's probably fine right now.

BobbyRBruce reviewed (commented):
Thanks Jason. I think this is really good. It gives me more confidence in our current design a you've not needed to alter too much to get this working. I've left some comments but nothing major. I enjoyed reading your notes :)

powerjg committed:
configs: Move some runtime checks in components

The runtime coherence check should check the cache heirarchy, not the
global protocol so it can differentiate between Ruby and classic.
This patch also removes some unnecessary includes and fails earlier in
MESI_Two_Level if that protocol isn't built.

Change-Id: I31c1fbd1d9597163b9138e80619d05c132a91545
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
misc,configs: Add gitignore for downloaded files

Change-Id: Idf9aa0e5e3b8490d1e75c8811a8dfad48bbdada3
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs: Update DTB generation for RISC-V board

This change moves the DTB generation into the board instead of using the
implicit DTB generation in many of the SimObjects. The problem with
using the implicit DTB generation is that it assumes things are named a
specific way (e.g., system.cpu). Since the component library doesn't
work that way, we need to reimplement the DTB generation. This has the
added benefit that the DTB generation can be specialized for RISC-V.

Change-Id: Ia018a812dd44a977bd41ceb5b9fed0228ff3614c
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
WIP,configs: Update component API for memory size

This change updates the API in the component library for setting the
size of memory. Now, you can set the size of the memory system as an
argument to the memory object. Then, the board is responsible for
figuring out what the overall memory ranges should be which it
communicates back to the memory system.

This should make multi-channel memories easier to implement and it fixes
some confusion around things like the HiFive platform starting at
0x8000000.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs: Update RISC-V board with new memory

Change-Id: I9b71e425a0c90cd0ed1c5d9913fee11141e60fb3
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

BobbyRBruce commented on components_library/memory/abstract_memory_system.py:
```diff
@@ -41,17 +41,33 @@ def __init__(self) -> None:
         raise NotImplementedError
 
     @abstractmethod
-    def get_memory_ranges(self) -> List[AddrRange]:
```
Surely we still want to be able to get the memory ranges?

powerjg replied:
Why? I would argue that should be hidden since the board should be in charge of this.

BobbyRBruce replied:
Fair point. I was battling over what should be in charge of memory ranges, but yes, the board is the place for it to be.

BobbyRBruce commented on components_library/boards/x86_board.py:
```diff
@@ -94,10 +95,17 @@ def __init__(
         self.mem_ranges = [
-            AddrRange(Addr("3GB")),  # All data
+            data_range,  # All data
             AddrRange(0xC0000000, size=0x100000),  # For I/0
```
There's nothing stopping these ranges overlapping right now, right?

powerjg replied:
There's a check to make sure data_range is < 3GB. That's the best we can do right now.

BobbyRBruce commented on components_library/memory/abstract_memory_system.py:
```diff
@@ -41,17 +41,33 @@ def __init__(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_memory_range(self, ranges: List[AddrRange]) -> None:
```
I see why you've done it this way, but I deeply dislike having to set the memory ranges in this way. At the core of my design is that all the board-component interaction should happen with the "incorporate_x" functions. Do you think there's a way you could migrate that there? Even then I'm not super happy with that as it's something someone might forget about :/

A more complex design to enforce this could be:

```
def incorporate_memory(...):
   _incorporate_memory(...)
   _set_memory_range(...)

@abstractmemory
def _incorporate_memory(...):
   pass

@abstractmethod
def _set_memory_range(...)
   pass
```

in the AbstractMemorySystem.

Just a thought, willing to discuss this.

powerjg replied:
Done!

powerjg committed:
configs: Remove unneeded ddr3 memory component

Change-Id: Id0228b4628c1be6dce3d7092731798bfa07de815
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

BobbyRBruce commented on components_library/boards/test_board.py:
```diff
@@ -61,12 +68,12 @@ def __init__(
     def connect_system_port(self, port: Port) -> None:
         self.system_port = port
 
     def connect_things(self) -> None:
```
Is there a good reason to not move this implementation to the `AbstractBoard`? This seems like pretty default behavior (I know X86 will need overridden regardless).

powerjg replied:
Yeah, you're right. That was a step I forgot to take.

BobbyRBruce commented on components_library/boards/abstract_board.py:
```diff
@@ -167,6 +165,26 @@ def set_mem_mode(self, mem_mode: MemMode) -> None:
+        # The simple board just has one memory range that is the size of the
+        # memory.
+        self.mem_ranges = [AddrRange(memory.get_size())]
+        memory.set_memory_range(self.mem_ranges)
```
I think, given your documentation here, you meant this to go in the Simple board, and this function to be abstract? That's how i'd do it.

powerjg replied:
Wait... why would this go in simple board and connect_things go in abstract? They should both be in one or the other.

BobbyRBruce commented on components_library/boards/x86_board.py:
```diff
@@ -267,7 +268,25 @@ def assignISAInt(irq, apicPin):
+        self.setup_memory_ranges()
+
+        # Set up all of the I/O before we incorporate anything else.
+        self._setup_io_devices()
```
This is quite annoying as it kind of breaks out design a little bit. I think a better way to do this would be to call `setup_memory_ranges()` in the `AbstractBoard` constructor. I know this is a _little bit_ ugly, but it allows us to develop new boards with the assumption that the memory ranges are already setup after construction of the base class. In this case, the `_setup_io_devices()` can be called in the constructor of the x86_board, and `connect_things()` can be kept simple.

powerjg replied:
I thought about that. But then you have to make sure to call super().__init__ at the *end* of the init function.

BobbyRBruce reviewed (commented):
Thanks Jason. I think this is really starting to come together and is strengthening our design. I've left some comments about moving stuff around a little.

powerjg commented:
Closing this PR since it's on gerrit now.
